File: viewer/src/main/scala/farjs/viewer/ViewerPluginUi.scala
package farjs.viewer

import farjs.filelist._
import farjs.ui.menu.{BottomMenu, BottomMenuProps}
import farjs.ui.popup.{Popup, PopupProps}
import farjs.viewer.ViewerPluginUi._
import scommons.react.blessed._
import scommons.react.hooks._
import scommons.react.{raw, _}

import scala.scalajs.js

class ViewerPluginUi(filePath: String, size: Double)
  extends FunctionComponent[FileListPluginUiProps] {

  protected def render(compProps: Props): ReactElement = {
    val inputRef = raw.React.useRef(null)
    val (viewport, setViewport) = useState(js.undefined: js.UndefOr[ViewerFileViewport])
    val props = compProps.plain

    val onKeypress: js.Function1[String, Boolean] = { keyFull =>
      var processed = true
      keyFull match {
        case "f3" | "f10" => props.onClose()
        case _ => processed = false
      }
      processed
    }

    val headerProps = viewport.toOption match {
      case None =>
        ViewerHeaderProps(filePath)
      case Some(vp) =>
        ViewerHeaderProps(
          filePath = filePath,
          encoding = vp.encoding,
          size = vp.size,
          column = vp.column,
          percent = vp.progress
        )
    }
    val menuItems = viewport.toOption match {
      case None => defaultMenuItems
      case Some(vp) =>
        var items = defaultMenuItems
        if (vp.wrap) {
          items = items.updated(1, "Unwrap")
        }
        items
    }

    <(popupComp)(^.plain := PopupProps(
      onClose = props.onClose,
      onKeypress = onKeypress
    ))(
      <.box(
        ^.rbClickable := true,
        ^.rbAutoFocus := false
      )(
        <(viewerHeader)(^.plain := headerProps)(),
  
        <.button(
          ^.ref := { el: BlessedElement =>
            inputRef.current = el
          },          ^.rbTop := 1,
          ^.rbWidth := "100%",
          ^.rbHeight := "100%-2"
        )(
          <(viewerController())(^.plain := ViewerControllerProps(
            inputRef = inputRef,
            dispatch = props.dispatch,
            filePath = filePath,
            size = size,
            viewport = viewport,
            setViewport = setViewport
          ))()
        ),
  
        <.box(^.rbTop := "100%-1")(
          <(bottomMenuComp)(^.plain := BottomMenuProps(menuItems))()
        )
      )
    )
  }
}

object ViewerPluginUi {

  private[viewer] var popupComp: ReactClass = Popup
  private[viewer] var viewerHeader: ReactClass = ViewerHeader
  private[viewer] var viewerController: UiComponent[ViewerControllerProps] = ViewerController
  private[viewer] var bottomMenuComp: ReactClass = BottomMenu

  private[viewer] val defaultMenuItems = js.Array(
    /*  F1 */ "",
    /*  F2 */ "Wrap",
    /*  F3 */ "Quit",
    /*  F4 */ "",
    /*  F5 */ "",
    /*  F6 */ "",
    /*  F7 */ "",
    /*  F8 */ "Encodings",
    /*  F9 */ "",
    /* F10 */ "Quit",
    /* F11 */ "",
    /* F12 */ "DevTools"
  )
}

File: viewer/ViewerPluginUi.mjs
/**
 * @typedef {import("@farjs/blessed").Widgets.BlessedElement} BlessedElement
 * @import { FileListPluginUiProps } from "@farjs/filelist/FileListPlugin.mjs";
 * @import { ViewerFileViewport } from "./ViewerFileViewport.mjs";
 */
import React, { useRef, useState } from "react";
import Popup from "@farjs/ui/popup/Popup.mjs";
import BottomMenu from "@farjs/ui/menu/BottomMenu.mjs";
import ViewerHeader from "./ViewerHeader.mjs";
import ViewerController from "./ViewerController.mjs";

const h = React.createElement;

/**
 * @param {string} filePath
 * @param {number} size
 */
function ViewerPluginUi(filePath, size) {
  /**
   * @param {FileListPluginUiProps} props
   */
  const ViewerPluginUiComp = (props) => {
    const {
      popupComp,
      viewerHeader,
      viewerController,
      bottomMenuComp,
      defaultMenuItems,
    } = ViewerPluginUi;

    const inputRef = /** @type {React.MutableRefObject<BlessedElement>} */ (
      useRef()
    );
    const [viewport, setViewport] = useState(
      /** @type {ViewerFileViewport | undefined} */ (undefined)
    );

    /** @type {(keyFull: string) => boolean} */
    function onKeypress(keyFull) {
      var processed = true;
      switch (keyFull) {
        case "f3":
        case "f10":
          props.onClose();
          break;
        default:
          processed = false;
          break;
      }
      return processed;
    }

    const menuItems = (() => {
      if (!viewport) {
        return defaultMenuItems;
      }

      const items = [...defaultMenuItems];
      if (viewport.wrap) {
        items[1] = "Unwrap";
      }
      return items;
    })();

    return h(
      popupComp,
      { onClose: props.onClose, onKeypress },
      h(
        "box",
        {
          clickable: true,
          autoFocus: false,
        },
        h(viewerHeader, {
          filePath: filePath,
          encoding: viewport?.encoding ?? "",
          size: viewport?.size ?? 0,
          column: viewport?.column ?? 0,
          percent: viewport?.progress ?? 0,
        }),

        h(
          "button",
          {
            ref: (/** @type {BlessedElement} */ el) => {
              inputRef.current = el;
            },
            top: 1,
            width: "100%",
            height: "100%-2",
          },
          h(viewerController, {
            inputRef,
            dispatch: props.dispatch,
            filePath,
            size,
            viewport,
            setViewport,
            onKeypress: () => false,
          })
        ),

        h("box", { top: "100%-1" }, h(bottomMenuComp, { items: menuItems }))
      )
    );
  };

  ViewerPluginUiComp.displayName = "ViewerPluginUi";

  return ViewerPluginUiComp;
}

ViewerPluginUi.popupComp = Popup;
ViewerPluginUi.viewerHeader = ViewerHeader;
ViewerPluginUi.viewerController = ViewerController;
ViewerPluginUi.bottomMenuComp = BottomMenu;

/** @type {readonly string[]} */
ViewerPluginUi.defaultMenuItems = [
  /*  F1 */ "",
  /*  F2 */ "Wrap",
  /*  F3 */ "Quit",
  /*  F4 */ "",
  /*  F5 */ "",
  /*  F6 */ "",
  /*  F7 */ "",
  /*  F8 */ "Encodings",
  /*  F9 */ "",
  /* F10 */ "Quit",
  /* F11 */ "",
  /* F12 */ "DevTools",
];

export default ViewerPluginUi;
